rfcs/0132-meson-builds-nix.md

```diff
@@ -0,0 +1,127 @@
+---
+feature: meson_builds_nix
+start-date: 2022-08-25
+author: Anderson Torres
+co-authors: @p01arst0rm
+shepherd-team:
+shepherd-leader:
+related-issues:
+---
+
+# Summary
+[summary]: #summary
+
+Use meson as an alternative build system for the reference implementation of Nix.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, the reference implementation of Nix evaluator, as well as its companion toolset generated from the Nix source tree, is built using the typical `./configure` shell script that relies on autoconf, along with standard GNU Make utility.
+
+Over time, this build system has been modified to keep up with the development needs of Nix projec. However, it has reached a state where the build system became clunky and plastered, hard to understand and modify, consequently making improvements to the project as a whole very difficult.
```

Comment:
I don't really buy this argument. The current build system is (IMHO) a pretty straight-forward use of Make and I have absolutely no indication that the build system is making Nix improvements "very difficult" - in general there isn't a lot of churn to the build system. Do you have examples of things that were made difficult because of the build system?

It would also need to be shown that a Meson build system would be easier to maintain. Certainly switching from a build system that I know to one that I don't know is not going to make *my* life easier, which is a problem since I'm the maintainer.

I'm in favor of getting rid of Autoconf. There is not a lot of value it provides these days. Most of the platforms checks can become `#if`s in the C++ code, and most of the library dependencies can be handled by calling `pkg-config` from the makefile.

Reply:
@edolstra if you are only assessing nix through the lens of gnu/linux and nixOS, then no, switching away from make/autotools will only result in minor improvements. the improvements only show themselves when you try and compile the autotools build on non gnu/linux platforms. meson & ninja allow significantly better support for these environments. a number of improvements i can think of right now include:

- better support for static/dynamic linking on non gnu/linux platforms
- better dependency inclusion with dynamic dependency objects. e.g. fixes issues on freebsd where hardcoded compiler flags differ between operating systems; previously you would need to add or change hardcoded flags stored in the makefiles.
- better support for non GCC compilers. Allows users to user XCode/Clang/MSVC to compile nix instead of having to locally create & maintain an entire userland & compiler with which to compile the nix project
- better support for nix unit testing (GTest, Valgrind)
- less complicated syntax
- removal of some build dependencies
- far better dependency discovery & inclusion with pkgconfig/cmake, for example often autotools will fail to find libreadline on Arch Linux, or LibZ on freebsd.
- better support for locating external system headers

Reply:
> Most of the platforms checks can become #ifs in the C++ code, and most of the library dependencies can be handled by calling pkg-config from the makefile.

This is also not a good idea. This would be a maintainability nightmare.

@edolstra it boils down to if we just care about Linux and MacOS, the current system is fine, but if we want to keep on expanding where we can run, it will not scale.

You can't empirically discover "it will not scale" from current experience on mainly those two platforms, you can only empirically discover it from trying to port it and finding the process sucks.

Reply:
> Most of the platforms checks can become #ifs in the C++ code,

This would be independently interesting to see implemented, as it would result in both autoconf and Meson running faster during the setup stage.

However, I suspect it often cannot be done, because the platform checks are either too complex to fit into a preprocessor, or they are used to define the output rules and mostly consist of checking the HOST_OS in configure, then repeatedly evaluating large switch statements in pure Make.

(I would not call that "a pretty straight-forward use of Make", nor would I call an umpteen-layers-deep call stack of `$(eval $(call ...))` such.)

> most of the library dependencies can be handled by calling pkg-config from the makefile

Repeatedly running pkg-config on every build sounds non-ideal, whatever else happens.

Reply:
> It would also need to be shown that a Meson build system would be easier to maintain.

I have included two references, one from Meson and other from NetBSD, showing differences between them.

Also I believe anything besides TECO is easier to understand than autotools' M4 and its generated files. 
And further, anything besides Python or Lua is harder to understand than Meson's DSL :)

> Most of the platforms checks can become #ifs in the C++ code

As the others said before, it can become a maintenance nightmare. Further I am of the team that believes C preprocessor should be treated as a necessary evil, to be avoided always when possible.

> and most of the library dependencies can be handled by calling pkg-config from the makefile.

Meson can handle them well - p01arst0rm argued this at length.

Reply:
The C preprocessor does excel at activating code blocks based on:
- announced software version numbers of dependencies in a header
- operating system name

Neither of which require feature macros in some kind of config.h

Also `cxx.has_header()` and a feature macro (Meson) could be replaced by `__has_include()` (C++17 standard).

I think these are all pretty reasonable, the problem is that actual project configurations generally need a lot more than that, which usually involves testing complex code snippets, or modifying the  linker command line.